cmd/end2endtest/zkweighted.go

```diff
@@ -163,8 +167,64 @@ func (t *E2EAnonElectionTempSIKs) Run() error {
 		return err
 	}
 
+	// the half of accounts should not have valid sik, due they are using temporal sik
+	halfAccts := len(votes) / 2
+	accs, err := removedSiks(t.api, votes, t.config.parallelCount)
+	if err != nil {
+		return err
+	}
+
+	if int(accs) != halfAccts {
+		return fmt.Errorf("unexpected number of accounts without valid SIK, got %d want %d", accs, halfAccts)
+	}
+
 	log.Infof("election %s status is RESULTS", t.election.ElectionID.String())
 	log.Infof("election results: %v", elres.Results)
 
 	return nil
 }
+
+// removedSiks get the number of accounts without a valid SIK
+func removedSiks(api *apiclient.HTTPclient, votes []*apiclient.VoteData, parallelCount int) (accts int32, err error) {
+	var siksRemoved int32
+	var wg sync.WaitGroup
+	errorChan := make(chan error, len(votes))
+
+	// worker pool channel with a maximum of parallelCount workers
+	workerPool := make(chan struct{}, parallelCount)
+
+	for _, v := range votes {
+		workerPool <- struct{}{} // reserves a slot in the worker pool
+		wg.Add(1)
+		go func(v *apiclient.VoteData) {
+			defer wg.Done()
+			defer func() { <-workerPool }() // make available for another goroutine to use
```

Comment:
i believe you should first do wg.Add(1), then reserve the slot, otherwise the wg.Wait could conceptually happen before you expect it to. in practice i think this will never happen, but if you reorder then AFAIU it can definitely never happen, conceptually

the order of defers is OK because the defer stack is LIFO, so the <-workerPool will happen before wg.Done (which is perfect)

Reply:
Sounds good, agree. I updated the code to reserve the slot after the wg.Add(1)